server/ServerApp.cpp

```diff
@@ -4538,9 +4538,15 @@ void ServerApp::PostCombatProcessTurns() {
     // UniverseObjects will have effects applied to them this turn, allowing
     // (for example) ships to have max fuel meters greater than 0 on the turn
     // they are created.
-    m_universe.ApplyMeterEffectsAndUpdateMeters(context, false);
+    std::vector<int> newly_created_ids;
+    for (auto& object : m_universe.Objects().all<UniverseObject>()) {
+        if (context.current_turn == object->CreationTurn())
+            newly_created_ids.push_back(object->ID());
+    }
+    if (!newly_created_ids.empty())
+        m_universe.ApplyMeterEffectsAndUpdateMeters(newly_created_ids, context, false);
```

Comment:
Could restricting the updated objects cause issues if the applicable effects use conditons that do things like limit the number of target objects, or apply an effect that changes in size depending on the number of affects objects?

Reply:
nice special cases ;)

should add them to test cases. I dont think one can query the targets directly or use the count of targets, but i am not sure what is inside the context used by the counting etc. conditions. 

> if the applicable effects use conditons that do things like limit the number of target objects, 

those will probably do weird/wrong stuff without the restriction too, if those have sitreps or other non-meter effects (meter and non-meter effects suddenly having different targets). Those effects should add something like created turn > current turn condition in the scope.

>or apply an effect that changes in size depending on the number of affects objects?

i guess the effects for old and newly created objects would differ, so again weird stuff if there are sitreps or non-meter effects. i guess usually the effects should only consider old objects when counting or only target old objects.

Reply:
> I dont think one can query the targets directly or use the count of targets, but i am not sure what is inside the context used by the counting etc. conditions.

You can match objects if (the number of objects that match another condition is within a range) using Number. Eg. "If there are 5 ships with shields over 10 in the same system, destroy a planet".

> those will probably do weird/wrong stuff without the restriction too, if those have sitreps or other non-meter effects (meter and non-meter effects suddenly having different targets).

Not especially if there's no randomness involved... Like "add a total of 5 to the max shield of ships in the system, divided evenly over all those ships".

> i guess the effects for old and newly created objects would differ, so again weird stuff if there are sitreps or non-meter effects. i guess usually the effects should only consider old objects when counting or only target old objects.

Why specifically non-meter effects?

Reply:
> > I dont think one can query the targets directly or use the count of targets
> 
> You can match objects if (the number of objects that match another condition is within a range) using Number. Eg. "If there are 5 ships with shields over 10 in the same system, destroy a planet".

Destruction is not a meter effect, so it would not trigger after production (and it would be based on the count of old ships). So that effect is not affected by this PR at all.

But I guess `Number condition = Target.Shield` would check the count of candidates, which with this PR would probably only include newly created ones. 

> 
> > those will probably do weird/wrong stuff without the restriction too, if those have sitreps or other non-meter effects (meter and non-meter effects suddenly having different targets).
> 
> Not especially if there's no randomness involved... Like "add a total of 5 to the max shield of ships in the system, divided evenly over all those ships".

I disagree. Its not only about randomness. Any difference in meters might be observable. E.g. if there are five ships in a system and another five get created, those ten ships would get each 0.5 extra max shields, but the sitrep would rather read for the old five ships  'Added 1 shield' and one would not get any sitreps for the newly created ones.

> 
> > i guess the effects for old and newly created objects would differ, so again weird stuff if there are sitreps or non-meter effects. i guess usually the effects should only consider old objects when counting or only target old objects.
> 
> Why specifically non-meter effects?

because non-meter effest get applied based on pre-production universe state. Meter effects get applied based on pre-production state, then are applied again based on post-production state.

Reply:
@geoffthemedio i guess we could make this slightly more correct by keeping all candidates and just not executing the effect on old the targets. Then in many corner cases I expect the target set after production would be more similar to the target set.

As example if scope is all ships (and there were O old ships and N new ships), an effect counting the number of targets would in post-production count N with this PR; with that change the count would be O+N. The count would be O pre-production. With O usually being much higher than N, the result would be much better.
I think for every default measure we take, we can construct an example which breaks it (in the sense there is a mismatch). 

So what are the options:

- go with something which is slightly incorrect for some (hypothetical) content
- go with something which is slightly less incorrect for some (hypothetical) content
- take one of the first two options and add some way to disable and override that default
- take one of the first two options and add some way to workaround that default (e.g. skip the effect for newly instantiated objects)
- instead of having a default which is based on using the normal meter effects, rely on explicit/special behavior for instantiation (either by a ~phase-filtering condition)

all of that options need some kind of clear documentation to FOCS scripters i guess 

the downside of explicit instantiation is that usually all meter effects should be applied and that very few effects should do something different for old vs new objects. For a filtering condition this would mean in every effect sth like: if in pre-production, apply the meter changes to all (i.e. all old) objects; and if in post-production, apply the meter changes to all new objects. Adding this everywhere seems excessive and ugly.

Reply:
> > Not especially if there's no randomness involved... Like "add a total of 5 to the max shield of ships in the system, divided evenly over all those ships".
> 
> I disagree. Its not only about randomness. Any difference in meters might be observable. E.g. if there are five ships in a system and another five get created, those ten ships would get each 0.5 extra max shields, but the sitrep would rather read for the old five ships 'Added 1 shield' and one would not get any sitreps for the newly created ones.

I wasn't thinking about sitreps, but I think we agree? If there is a non-random test based on number of ships that is given a restricted universe subset as candidates and also uses that to do counts of the numbers of objects, it would miscount. I'm not sure if it uses the potential candidates subset for such condition evaluations, though. More likely to be inconsistent would be conditions that match a certain number of objects or do a ranking of objects by some value, like the top 5 highest structure ships. Those I'm more sure would/could give inconsistent results with a separate post production restricted candidates meter update.

Reply:
> but I think we agree? If there is a non-random test based on number of ships that is given a restricted universe subset as candidates and also uses that to do counts of the numbers of objects, it would miscount.

yes agreed. 

> I'm not sure if it uses the potential candidates subset for such condition evaluations, though. More likely to be inconsistent would be conditions that match a certain number of objects or do a ranking of objects by some value, like the top 5 highest structure ships. 

actually i think it takes the context universe for evaluating such sub-conditions like in statistics and also Number and NumberOf. Actually currently I cant think of a way of filtering such a condition by the candidate set. Can you? If not I guess _this_ problem actually does not exist? will just play test to be sure.


> To rephrase, both what objects are selected by a condition and what modification to apply can depend on counts or rankings if objects. If just picking from a subset, then the rankings could be distorted...
> Depending on the specific effect and condition, restricting the root candidates to new objects or old objects or allowing all objects could be correct, and similarly restricting candidates for subconditions and refs could be correct in the sense of what should be matched and consistency with sitrep generation...

yes. But even if have full control of restricting or not restricting the target set  sometimes an effect could still be impossible. like destroying the 5 highest structure ships including new ships. or boosting the 5 lowest structure ship including new ships and giving the player a sitrep about it. But with enough control, one can exclude new ships from such effects.

I just figured that newly created ships should have zero structure until very late in effects processing? If so i could use this to differentiate old and new ships in conditions (those with initial structure bigger zero are old ships) - I could not make comparing creation turn with current turn work for that purpose.

So revisiting the effect phase idea: if we would split pre-production and post-production effects completely, the post-production effects could contain non-meter effects and thus e.g. sitreps could be consistent with post-production. Control wise this is certainly the most straightforward thing to do. maybe there would be a big subset of effects which would only happen post-production; with current contents only the effects for population, influence?, industry and research would need to happen before production. all the effects setting ship meters (besides industry and research) could happen after production and automatically apply to the full set of (old and new) ships. 
Of course effects which would mix such meters (e.g. extra industry based on military strength) would then become prone to have +1 turn effects or would need to be carefully scripted for both phases.

What this means for estimating effective meter values (also for showing in the GUI).. I dont know. Maybe showing meters for certain phases makes sense: damage meters for pre-combat phase; research and industry should be shown pre-production. not sure when influence is spend, the significant value to show is the meter value directly beforehand.

But if we are that far actually the more KISS solution is to get rid of pre-production; just move all effects to post-production. We would have a +1 turn effect on production meter changes. But one would be certain that one would be able to spent exactly what the meters were at the end of the last turn (which is simple and direct) - minus production from lost resource centers.

Reply:
> > I'm not sure if it uses the potential candidates subset for such condition evaluations, though. More likely to be inconsistent would be conditions that match a certain number of objects or do a ranking of objects by some value, like the top 5 highest structure ships.
> 
> actually i think it takes the context universe for evaluating such sub-conditions like in statistics and also Number and NumberOf.

Sub-condition evaluation normally use the whole universe (or at least the objects in the ScriptingContext being used). So, for exmaple, `Number low = 5 condition = Ship` will evaluate the `Ship` subcondition on all objects in the universe, regardless of what objects are being matched or considered as potential matches by the `Number` condition. For this type of condition, the sub-condition matches are a separate set of objects from the main condition matches. `NumberOf` is different, in that it applies the sub-condition to filter the main condition candidate objects before selecting the specified number of them, rather than using a separate sub-condition evalulation as part of deciding which objects match the main condition like "Contains".

> Actually currently I cant think of a way of filtering such a condition by the candidate set. Can you? If not I guess _this_ problem actually does not exist? will just play test to be sure.

Not sure what you meaning by "filtering such a condition by the candidate set". Maybe if the subcondition includes a reference to `RootCandidate`? That would cause the subcondition to be separately evaluated for each candidate object for the top-level condition being evaluated. For example, testing if something is within some distance of another object that further has some restrictions like being owned by the same empire. The subcondition matches will be different for objects (root candidates) that have different owners.

> edit: even with this PR i can get the highest object id of all old ships in post production (on a newly built ship); this proves that the candidates in the Statistic max condition are not restricted.

Yes, that's as expected.

> edit: actually `(LocalCandidate.CreationTurn < CurrentTurn)` is working for finding out if the object is old (and it evaluates to true in preproduction as well).

That's what I'd use to select just new or just old ships with respect to production. Sorry if that wasn't clear... I didn't think I needed to explain how to do that.

> initial structure seems never to be zero not sure how that works.

All paired meter values are set to high values when new ships are created, so that later they can be clamped by the post-effect max meter values of those ships. see: https://github.com/freeorion/freeorion/blob/master/Empire/Empire.cpp#L2660

Reply:
> move all effects to post-production. We would have a +1 turn effect on production meter changes. But one would be certain that one would be able to spent exactly what the meters were at the end of the last turn (which is simple and direct) - minus production from lost resource centers.

That would probably mean that focus changes and new buildings or new techs would not be able to affect production output until a turn later...

Reply:
> > move all effects to post-production. We would have a +1 turn effect on production meter changes. But one would be certain that one would be able to spent exactly what the meters were at the end of the last turn (which is simple and direct) - minus production from lost resource centers.
> 
> That would probably mean that focus changes and new buildings or new techs would not be able to affect production output until a turn later...

yes, i am certain of this. `We would have a +1 turn effect on production meter changes.`
But why it seems you think this is a problem?

Reply:
My vague recollection is that people have previously complained about buildings taking an extra turn to start working. For focus, it just seems like a better form of feedback to the player if some sort of feedback is immediately visible. For policies, it's also helpful in cases where a change could produce a big difference in eg. outputs of planets to at least see some negative effects immediately. But for focus and policies, I guess that's not really relevant to a discussion of pre/post production effects.

Reply:
> My vague recollection is that people have previously complained...

i can do a poll to get some feedback

> But for focus and policies, I guess that's not really relevant to a discussion of pre/post production effects.

hm. yes, probably you are right. If I switch e.g. to industry focus the immediate change would be that max industry and industry will increase next turn, not that the that industry is already available for building ships or buildings.